apps/rewards/src/invite/rewardSender.service.ts

```diff
@@ -0,0 +1,203 @@
+import { InviteRewardRepository } from './inviteReward.repository'
+
+const WEI_PER_UNIT = 1000000000000000000
+const MAX_INVITES_PER_QUERY = 100
+const REWARD_VALUE_CACHE_DURATION = 1 * 60 * 1000 // One minute
+
+interface WatchedInvite {
+  inviteId: string
+  txHash: string
+  sentAt: number
+  markAsDeadLetterIfFailed: boolean
+}
+
+@Injectable()
+export class RewardSenderService {
+  watchedInvites: Set<WatchedInvite>
+  rewardInCelo = {
+    reward: '',
+    timestamp: 0
+  }
+  checkingWatchedInvites = false
+  checkingFailedInvites = false
+
+  constructor(
+    private readonly inviteRewardRepository: InviteRewardRepository,
+    private readonly relayerProxyService: RelayerProxyService,
+    @Inject(appConfig.KEY)
+    private readonly appCfg: AppConfig,
+    private readonly contractKit: ContractKit,
+    private readonly logger: KomenciLoggerService
+  ) {
+    this.watchedInvites = new Set()
+  }
+
+  async wasTxSentSuccesfully(txHash: string) {
+    const tx = await this.contractKit.web3.eth.getTransaction(txHash)
+    return Boolean(tx.blockHash)
+  }
+
+  @Cron(CronExpression.EVERY_SECOND)
+  async checkWatchedInvites() {
+    if (this.checkingWatchedInvites) {
+      this.logger.debug(
+        'Skipping checking watched invites because previous run is still running'
+      )
+      return
+    }
+    this.checkingWatchedInvites = true
+    try {
+      await Promise.allSettled(
+        [...this.watchedInvites].map(async invite => {
+          if (await this.wasTxSentSuccesfully(invite.txHash)) {
+            await this.inviteRewardRepository.update(invite.inviteId, {
+              state: RewardStatus.Completed
+            })
+            this.watchedInvites.delete(invite)
+          } else if (
+            Date.now() - invite.sentAt >
+            this.appCfg.transactionTimeoutMs
+          ) {
+            await this.inviteRewardRepository.update(invite.inviteId, {
+              state: invite.markAsDeadLetterIfFailed
+                ? RewardStatus.DeadLettered
+                : RewardStatus.Failed
+            })
+            this.watchedInvites.delete(invite)
+          }
+        })
+      )
+    } catch (error) {
+      this.logger.error(error, 'Error checking watched invites')
+    } finally {
+      this.checkingWatchedInvites = false
+    }
+  }
+
+  @Cron(CronExpression.EVERY_MINUTE)
+  async checkFailedInvites() {
+    if (this.checkingFailedInvites) {
+      this.logger.debug(
+        'Skipping checking failed invites because the previous run still running'
+      )
+      return
+    }
+    this.checkingFailedInvites = true
+    try {
+      await this.inviteRewardRepository.manager.transaction(
+        async (entityManager: EntityManager) => {
+          const repository = entityManager.getRepository<InviteReward>(
+            'invite_reward'
+          )
+          const invites = await repository
+            .createQueryBuilder()
+            .setLock('pessimistic_read')
```

Comment:
This should be pessimistic_write, correct? We don't want two reward service instances reading this entity and deciding to retry the payment

Reply:
Yeah, I thought I had tested this with 2 instances but looks like I tested it wrong. I'll try again to find a good way to test this and change to `pessimistic_write`